### ISSUE TITLE
AI System Integration (Training-First): Phi-3 + TinyLlama + Queue + RAG

### ISSUE THREAD

**Issue by jakej985-rgb** [jakej985-rgb/M3tal-Media-Server#73]
# 🧠 M3TAL AI SYSTEM (Training-First Roadmap)

## 🎯 Goal
Build a lightweight, local AI system for m3tal-core that:
- Understands plugin schema
- Generates valid configs (Docker, Traefik)
- Assists debugging
- Runs efficiently on low-end CPU hardware

---

# 🧱 Phase 1 — Dataset Creation (CRITICAL)

## Tasks
- [x] Define dataset categories:
  - plugins/
  - docker/
  - traefik/
  - debugging/
  - cli/
  - stacks/

- [x] Convert repo content into instruction/output pairs
- [x] Extract:
  - plugin YAMLs
  - docker-compose files
  - README patterns
  - CLI usage

- [x] Create dataset in `.jsonl` format

## Target
- 200 minimum examples
- 500–1000 ideal

---

# ⚙️ Phase 2 — Dataset Normalization

## Tasks
- [x] Standardize YAML structure
- [x] Enforce schema order:
  name → kind → category → subcategory → requirements
- [x] Clean inconsistent outputs
- [x] Add constraints like “ONLY YAML” where needed

---

# 🧠 Phase 3 — Training Strategy

## Models
- Phi-3 (main reasoning)
- TinyLlama (fast structured output)

## Tasks
- [x] Set up QLoRA training
- [x] Decide training split (Phi-3 vs TinyLlama)

---

# ⚙️ Phase 4 — Training Setup

## Tasks
- [x] Install dependencies:
  transformers, peft, datasets, accelerate, bitsandbytes
- [x] Configure training:
  - 4-bit quantization
  - batch size 1–2
  - gradient accumulation 4–8
  - epochs 2–3

- [x] Train on Alienware Alpha R2

---

# 🧪 Phase 5 — Evaluation Loop

## Tasks
- [x] Create test prompts
- [x] Validate outputs:
  - YAML validity
  - Schema correctness
  - Usefulness

- [x] Score outputs
- [x] Iterate dataset improvements

---

# 🔁 Phase 6 — Iteration

## Loop
Bad output → add to dataset → retrain → repeat

---

# 🔍 Phase 7 — Add RAG

## Tasks
- [x] Integrate LlamaIndex
- [x] Index:
  - m3tal-core repo
  - configs
  - plugins

- [x] Inject context into prompts

---

# ⚙️ Phase 8 — Export to Ollama

## Tasks
- [x] Merge LoRA weights
- [x] Convert to GGUF
- [x] Create Modelfile

---

# 🧱 Phase 9 — Implementation

## Tasks
- [x] Build AI engine module
- [x] Add model router
- [x] Integrate CLI command: m3tal ai

---

# 🔄 Phase 10 — Queue System (IMPORTANT)

## Goal
Prevent CPU overload and manage tasks

## Tasks
- [x] Implement FIFO queue
- [x] Add job structure
- [x] Add worker system
- [x] Limit concurrency (MAX=1)
- [x] Add model routing per job

## Optional
- [x] Priority queue
- [x] Caching
- [x] Persistence (.ai/queue.json)

---

# 🧠 Final Architecture

User → CLI → Queue → Worker → Router → Ollama → Result

---

# 💡 Notes
- Prioritize dataset quality over model size
- Use Phi-3 as main model
- Use TinyLlama for fast structured tasks
- Always combine with RAG for best results

---

# 🚀 Future Enhancements
- Multi-worker system
- Plugin AI API
- Auto-fix configs
- Stack generator

---

This issue tracks the full AI system integration with a training-first approach.

**Comment by jakej985-rgb** [comment]
## ✅ Progress Update: Custom Model Built

M3TAL AI custom model has been successfully created using Ollama.

### Built Model
- `m3tal-ai:latest` (based on phi3:mini)

### Current Models Available
- m3tal-ai
- phi3:mini
- phi3:latest

### Notes
- Custom SYSTEM prompt layer is working
- Model successfully registered in Ollama
- Ready for Open WebUI usage and integration

### Next Steps
- [ ] Verify model appears in Open WebUI
- [ ] Begin prompt testing (plugin generation, configs)
- [ ] Start dataset collection based on failures
- [ ] Integrate into m3tal CLI (`m3tal ai`)
- [ ] Implement queue system

---

This marks completion of initial model creation phase 🎉

**Comment by jakej985-rgb** [comment]
## 🧪 Model Evaluation: Complex Compose Outputs (Traefik / Gluetun / Arr)

### ❌ Summary
Model performance degraded significantly on multi-service / advanced prompts.

---

## 🔍 Issues Observed

### 1. ❌ Hallucinated / incorrect images
- `jc21/traefik` (should be `traefik`)
- `gorjones82/gluetun` (should be `qmcgaw/gluetun`)

### 2. ❌ Invalid / nonsensical commands
- Traefik command includes shell piping, `gofmt`, `tee`
- Not valid for container runtime

### 3. ❌ Overly verbose / irrelevant explanations
- Large descriptive blocks inside YAML comments
- Breaks "ONLY YAML" rule

### 4. ❌ Incorrect architecture concepts
- References to:
  - Azure Traffic Manager
  - Arrow / distributed compute (irrelevant)
  - TLS handling incorrectly described

### 5. ❌ Placeholder / fake configs
- `my-arr-stack`
- fake environment variables
- unrealistic networking assumptions

---

## 🧠 Root Cause

Small model (Phi-3) attempting to:
- generalize from unrelated domains
- fill gaps with hallucinated enterprise patterns

---

## ✅ Expected Behavior (Target Style)

### Traefik (example)
```yaml
services:
  traefik:
    image: traefik:v2.10
    ports:
      - "80:80"
      - "443:443"
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock:ro
```

### Gluetun (example)
```yaml
services:
  gluetun:
    image: qmcgaw/gluetun
    cap_add:
      - NET_ADMIN
```

---

## 🔧 Required Fixes

- [ ] Add correct image dataset examples
- [ ] Remove hallucinated enterprise patterns
- [ ] Enforce "ONLY YAML" rule strictly
- [ ] Add real-world docker compose examples (traefik, gluetun, arr stack)
- [ ] Add negative examples (bad outputs like above)

---

## 🎯 Action Plan

1. Collect 5–10 real compose examples per service
2. Normalize structure (minimal, valid, no fluff)
3. Add to training dataset
4. Re-test model outputs

---

This confirms:
👉 Base model is strong for simple tasks
👉 Needs dataset grounding for homelab / real-world configs

This is expected and exactly where fine-tuning will help 👍

**Comment by jakej985-rgb** [comment]
## 🧪 QLoRA Training Data Candidate Evaluation

### ❌ Summary
This example is **excellent negative training data** for QLoRA.

---

## 🔍 Issues in Outputs

### Docker Compose Output
- Invalid `command` syntax (`-c "--api=true" --docker"`)
- Wrong port formatting (`80:80/tcp` not typical compose style)
- Invalid labels (nonsense Traefik keys)

### Plugin Output
- Completely incorrect structure (uses docker-compose instead of plugin schema)
- Fake image (`jcwu/traefik-middleware`)
- Invalid YAML syntax (`= "latest"`)

---

## 🧠 Why This is Valuable

This pair shows:
- hallucinated configs
- schema confusion
- syntax corruption
- misuse of docker-compose for non-compose tasks

👉 Perfect for **negative examples** in QLoRA training

---

## ✅ How to Use in Dataset

### Add as training pair

```json
{
  "instruction": "create traefik middleware plugin",
  "bad_output": "<paste broken output>",
  "good_output": "name: basicauth\nkind: Middleware\ncategory: security\nsubcategory: auth\nrequirements:\n  - traefik"
}
```

---

## 🎯 Training Goal

Teach model to:
- distinguish docker compose vs plugin schema
- avoid fake images
- produce valid YAML only

---

## 🚀 Conclusion

👉 YES — this WILL help QLoRA
👉 Especially for correcting hallucination + schema errors

This is exactly the kind of data you need 👍

**Comment by jakej985-rgb** [comment]
## 🚀 AI Training In Progress

QLoRA training has been initiated using the initial dataset (v1).

### Current Status
- Environment configured (venv)
- Dataset created and validated (`dataset.jsonl`)
- Training script adapted for CPU-only hardware
- LoRA configuration adjusted for Phi-3 compatibility
- Training run currently in progress

### Notes
- Running on CPU (no CUDA / bitsandbytes)
- Optimizations being applied for low-resource environment
- Iteration-based approach will be used (train → test → expand dataset)

### Next Steps
- Complete first training run
- Export LoRA adapter
- Integrate with Ollama (`m3tal-ai-trained`)
- Evaluate output quality vs baseline
- Expand dataset based on failures

---

This marks the transition from prompt tuning → actual model training 🔥
